music21/osmd/osmd.py

```diff
@@ -0,0 +1,222 @@
+            self.display(self.Javascript(script))
+        else:
+
+            tmp = tempfile.NamedTemporaryFile(delete=False)
+            tmp_path = tmp.name + '.html'
+
+            open(tmp_path, 'w').write(f"""
+            <div id="{divId}"></div>
+            <script>{script}</script>
+            """)
+            filename = 'file:///' + tmp_path
+            webbrowser.open_new_tab(filename)
+        return divId
+
+    def parseData(self):
+        raise NotImplementedError('osmd is display-only')
+
+    @staticmethod
+    def getUniqueDivId():
+        '''
+        Generates a unique id for the div in which the score is displayed.
+        This is so we can update a previously used div.
+        '''
+        return ("OSMD-div-"
+                + str(random.randint(0, 1000000))
+                + "-" + str(time.time()).replace('.', '-'))  # '.' is the class selector
+
+    def musicXMLToScript(self, xml, divId, offline=False):
+        '''
+        Converts the xml into Javascript which can be injected into a webpage to display the score.
+        If divId is provided then it will be used as the container, if not a new
+        '''
+        # script that will replace div contents with OSMD display
+
+        script_path = os.path.join(getSourceFilePath(), 'osmd', 'notebookOSMDLoader.js')
+        script = open(script_path, 'r').read() \
+            .replace('{{DIV_ID}}', divId) \
+            .replace('"{{data}}"', json.dumps(xml))
+
+        if offline is True:
+            if not os.path.isfile(self.osmd_file):
+                # on first use we download the file and store it locally for subsequent use
+                # TODO: catch error for no internet connection
+                urllib.request.urlretrieve(self.script_url, self.osmd_file)
+
+            # since we can't link to files from a notebook (security risk) inject into file.
+            script_content = open(self.osmd_file).read()
+            script = script.replace('"{{offline_script}}"', json.dumps(script_content), 1)
+
+        else:
+            script = script.replace('"{{script_url}}"', json.dumps(self.script_url), 1)
+        return script
+
+    @staticmethod
+    def addDefaultPartName(score):
+        '''
+        Due to a bug in OpenSheetMusicDisplay if a <part-name> is not provided it will default
+        to an auto-generated random string. This method is used to fix that by default.
+
+
+        >>> import music21
+        >>> # display two scores to demonstrate difference.
+        >>> s = music21.converter.parse("tinyNotation: 3/4 E4 r f# g=lastG trip{b-8 a g} c4~ c")
+        >>> _ = s.show('osmd',fixPartName=False)
+        >>> music21.osmd.ConverterOpenSheetMusicDisplay.addDefaultPartName(s)
+        >>> _ = s.show('osmd',fixPartName=False)
+        '''
+        # If no partName is present in the first instrument, OSMD will display the ugly 'partId'
+        allInstruments = list(score.getInstruments(returnDefault=False, recurse=True))
```

Comment:
Casting to list is unnecessary -- getInstruments returns a stream of elements, and you can use .first()